{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Title: Imputing missing gene expression values using 3D gene conformation data and kriging interpolation  \n",
    "使用三维基因构象数据和kriging插值填补缺失的基因表达值"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Background:  \n",
    "Genes are organized in 3D conformation in the nucleus, and genes closer in spatial proximity tend to be coregulated. While single-cell genomics techniques have profiled gene expression in individual cells, some values are often missing due to technical limitations.\n",
    "\n",
    "#### Objective:   \n",
    "The objective of this project is to leverage known population-level 3D gene conformations to impute missing gene expression values in single cells.\n",
    "\n",
    "#### Method:   \n",
    "We will first obtain the spatial covariance matrix of gene-gene distances using population-level Hi-C or other 3D genome data. For individual cells with partial expression measurements, we will apply kriging interpolation - a geostatistical technique using the spatial covariance matrix and known expression values to predict the missing values.\n",
    "\n",
    "#### Expected outcomes:  \n",
    "We expect this method can more accurately impute missing gene expression compared to alternative methods without using spatial information. It could help improve downstream analyses like cell clustering and trajectory inference from single cell expression data.\n",
    "\n",
    "The proposed method leverages 3D genome architecture to address a common challenge in single cell genomics data analysis. \n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 任务描述  \n",
    "将geography里的kriging interpolation应用到gene 3D conformation\n",
    "\n",
    "- 了解kriging interpolation以及相关的python包\n",
    "- 已知所有细胞共享的gene 3D conformation，获得gene-gene spatial covariance\n",
    "- 单个细胞中有部分的基因表达未知，使用covariance矩阵以及已知的基因表达值，impute未知基因表达值\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### kriging interpolation"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "- **基本假设与核心思想**\n",
    "  - 一点的属性值与其周围点的属性值有关，且可以由其周围点的属性值推导出\n",
    "\n",
    "- **地理学第一定律**\n",
    "  - 所有事物都与其他事物相关，但是近处的事物比远处的事物更相关\n",
    "\n",
    "- **克里金插值核心思想**\n",
    "  - 两点属性值差异性（不相关性）与二者间距离存在一定距离范围内成正相关"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Input"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Requirement already satisfied: anndata in c:\\users\\speranza\\appdata\\local\\programs\\python\\python310\\lib\\site-packages (0.10.7)\n",
      "Requirement already satisfied: numpy>=1.23 in c:\\users\\speranza\\appdata\\local\\programs\\python\\python310\\lib\\site-packages (from anndata) (1.26.4)\n",
      "Requirement already satisfied: pandas!=2.1.0rc0,!=2.1.2,>=1.4 in c:\\users\\speranza\\appdata\\local\\programs\\python\\python310\\lib\\site-packages (from anndata) (2.2.2)\n",
      "Requirement already satisfied: natsort in c:\\users\\speranza\\appdata\\local\\programs\\python\\python310\\lib\\site-packages (from anndata) (8.4.0)\n",
      "Requirement already satisfied: scipy>1.8 in c:\\users\\speranza\\appdata\\local\\programs\\python\\python310\\lib\\site-packages (from anndata) (1.13.1)\n",
      "Requirement already satisfied: exceptiongroup in c:\\users\\speranza\\appdata\\local\\programs\\python\\python310\\lib\\site-packages (from anndata) (1.2.0)\n",
      "Requirement already satisfied: h5py>=3.1 in c:\\users\\speranza\\appdata\\local\\programs\\python\\python310\\lib\\site-packages (from anndata) (3.11.0)\n",
      "Requirement already satisfied: packaging>=20.0 in c:\\users\\speranza\\appdata\\local\\programs\\python\\python310\\lib\\site-packages (from anndata) (23.2)\n",
      "Requirement already satisfied: array-api-compat!=1.5,>1.4 in c:\\users\\speranza\\appdata\\local\\programs\\python\\python310\\lib\\site-packages (from anndata) (1.7.1)\n",
      "Requirement already satisfied: tzdata>=2022.7 in c:\\users\\speranza\\appdata\\local\\programs\\python\\python310\\lib\\site-packages (from pandas!=2.1.0rc0,!=2.1.2,>=1.4->anndata) (2024.1)\n",
      "Requirement already satisfied: pytz>=2020.1 in c:\\users\\speranza\\appdata\\local\\programs\\python\\python310\\lib\\site-packages (from pandas!=2.1.0rc0,!=2.1.2,>=1.4->anndata) (2024.1)\n",
      "Requirement already satisfied: python-dateutil>=2.8.2 in c:\\users\\speranza\\appdata\\roaming\\python\\python310\\site-packages (from pandas!=2.1.0rc0,!=2.1.2,>=1.4->anndata) (2.8.2)\n",
      "Requirement already satisfied: six>=1.5 in c:\\users\\speranza\\appdata\\roaming\\python\\python310\\site-packages (from python-dateutil>=2.8.2->pandas!=2.1.0rc0,!=2.1.2,>=1.4->anndata) (1.16.0)\n",
      "Note: you may need to restart the kernel to use updated packages.\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "WARNING: You are using pip version 22.0.4; however, version 24.0 is available.\n",
      "You should consider upgrading via the 'c:\\Users\\Speranza\\AppData\\Local\\Programs\\Python\\Python310\\python.exe -m pip install --upgrade pip' command.\n"
     ]
    }
   ],
   "source": [
    "%pip install anndata"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Requirement already satisfied: h5py in c:\\users\\speranza\\appdata\\local\\programs\\python\\python310\\lib\\site-packages (3.11.0)\n",
      "Could not fetch URL https://pypi.org/simple/h5py/: There was a problem confirming the ssl certificate: HTTPSConnectionPool(host='pypi.org', port=443): Max retries exceeded with url: /simple/h5py/ (Caused by SSLError(SSLEOFError(8, 'EOF occurred in violation of protocol (_ssl.c:997)'))) - skipping\n",
      "Requirement already satisfied: numpy>=1.17.3 in c:\\users\\speranza\\appdata\\local\\programs\\python\\python310\\lib\\site-packages (from h5py) (1.26.4)\n",
      "Note: you may need to restart the kernel to use updated packages.\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "WARNING: Retrying (Retry(total=4, connect=None, read=None, redirect=None, status=None)) after connection broken by 'SSLError(SSLEOFError(8, 'EOF occurred in violation of protocol (_ssl.c:997)'))': /simple/h5py/\n",
      "WARNING: Retrying (Retry(total=3, connect=None, read=None, redirect=None, status=None)) after connection broken by 'SSLError(SSLEOFError(8, 'EOF occurred in violation of protocol (_ssl.c:997)'))': /simple/h5py/\n",
      "WARNING: Retrying (Retry(total=2, connect=None, read=None, redirect=None, status=None)) after connection broken by 'SSLError(SSLEOFError(8, 'EOF occurred in violation of protocol (_ssl.c:997)'))': /simple/h5py/\n",
      "WARNING: Retrying (Retry(total=1, connect=None, read=None, redirect=None, status=None)) after connection broken by 'SSLError(SSLEOFError(8, 'EOF occurred in violation of protocol (_ssl.c:997)'))': /simple/h5py/\n",
      "WARNING: Retrying (Retry(total=0, connect=None, read=None, redirect=None, status=None)) after connection broken by 'SSLError(SSLEOFError(8, 'EOF occurred in violation of protocol (_ssl.c:997)'))': /simple/h5py/\n",
      "WARNING: You are using pip version 22.0.4; however, version 24.0 is available.\n",
      "You should consider upgrading via the 'c:\\Users\\Speranza\\AppData\\Local\\Programs\\Python\\Python310\\python.exe -m pip install --upgrade pip' command.\n"
     ]
    }
   ],
   "source": [
    "%pip install --upgrade h5py"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "import anndata\n",
    "adata = anndata.read_h5ad('input\\\\adata.h5ad')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>LGALS3</th>\n",
       "      <th>SERPINA1</th>\n",
       "      <th>FAS</th>\n",
       "      <th>KITLG</th>\n",
       "      <th>VCAN</th>\n",
       "      <th>CD53</th>\n",
       "      <th>BMP4</th>\n",
       "      <th>NEAT1</th>\n",
       "      <th>CD63</th>\n",
       "      <th>LYZ</th>\n",
       "      <th>...</th>\n",
       "      <th>FABP4</th>\n",
       "      <th>CD69</th>\n",
       "      <th>NegPrb7</th>\n",
       "      <th>TNFRSF10A</th>\n",
       "      <th>NegPrb17</th>\n",
       "      <th>MMP12</th>\n",
       "      <th>FGF12</th>\n",
       "      <th>IFNA1</th>\n",
       "      <th>NRIP3</th>\n",
       "      <th>DPP4</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>c_2_1_311</th>\n",
       "      <td>7.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>4.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>5.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>11.0</td>\n",
       "      <td>7.0</td>\n",
       "      <td>26.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>c_2_1_383</th>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>6.0</td>\n",
       "      <td>5.0</td>\n",
       "      <td>5.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>c_2_1_513</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>6.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>7.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>c_2_1_525</th>\n",
       "      <td>2.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>5.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>14.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>13.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>c_2_1_589</th>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>9.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>c_2_30_3778</th>\n",
       "      <td>0.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>c_2_30_3808</th>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>5.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>c_2_30_3947</th>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>4.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>c_2_30_3987</th>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>c_2_30_4092</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1767 rows × 980 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "             LGALS3  SERPINA1  FAS  KITLG  VCAN  CD53  BMP4  NEAT1  CD63  \\\n",
       "c_2_1_311       7.0       1.0  4.0    1.0   3.0   5.0   1.0   11.0   7.0   \n",
       "c_2_1_383       1.0       0.0  0.0    0.0   0.0   0.0   0.0    6.0   5.0   \n",
       "c_2_1_513       0.0       0.0  0.0    1.0   3.0   1.0   0.0    6.0   1.0   \n",
       "c_2_1_525       2.0       2.0  0.0    0.0   5.0   0.0   0.0   14.0   0.0   \n",
       "c_2_1_589       1.0       1.0  0.0    0.0   3.0   0.0   0.0    1.0   2.0   \n",
       "...             ...       ...  ...    ...   ...   ...   ...    ...   ...   \n",
       "c_2_30_3778     0.0       2.0  0.0    0.0   1.0   0.0   0.0    0.0   0.0   \n",
       "c_2_30_3808     0.0       1.0  0.0    0.0   2.0   2.0   0.0    3.0   1.0   \n",
       "c_2_30_3947     0.0       1.0  0.0    0.0   0.0   1.0   0.0    0.0   0.0   \n",
       "c_2_30_3987     1.0       1.0  0.0    0.0   2.0   1.0   0.0    2.0   1.0   \n",
       "c_2_30_4092     0.0       0.0  0.0    0.0   0.0   0.0   0.0    1.0   1.0   \n",
       "\n",
       "              LYZ  ...  FABP4  CD69  NegPrb7  TNFRSF10A  NegPrb17  MMP12  \\\n",
       "c_2_1_311    26.0  ...    0.0   0.0      0.0        0.0       0.0    0.0   \n",
       "c_2_1_383     5.0  ...    0.0   0.0      0.0        0.0       0.0    0.0   \n",
       "c_2_1_513     7.0  ...    0.0   0.0      0.0        0.0       0.0    0.0   \n",
       "c_2_1_525    13.0  ...    0.0   0.0      0.0        0.0       0.0    0.0   \n",
       "c_2_1_589     9.0  ...    0.0   0.0      0.0        0.0       0.0    0.0   \n",
       "...           ...  ...    ...   ...      ...        ...       ...    ...   \n",
       "c_2_30_3778   0.0  ...    0.0   0.0      0.0        0.0       0.0    0.0   \n",
       "c_2_30_3808   5.0  ...    0.0   0.0      0.0        0.0       0.0    0.0   \n",
       "c_2_30_3947   4.0  ...    0.0   1.0      0.0        0.0       0.0    0.0   \n",
       "c_2_30_3987   2.0  ...    0.0   0.0      0.0        0.0       0.0    1.0   \n",
       "c_2_30_4092   2.0  ...    0.0   0.0      0.0        0.0       0.0    0.0   \n",
       "\n",
       "             FGF12  IFNA1  NRIP3  DPP4  \n",
       "c_2_1_311      0.0    0.0    0.0   0.0  \n",
       "c_2_1_383      0.0    0.0    0.0   0.0  \n",
       "c_2_1_513      0.0    0.0    0.0   0.0  \n",
       "c_2_1_525      0.0    0.0    0.0   0.0  \n",
       "c_2_1_589      0.0    0.0    0.0   0.0  \n",
       "...            ...    ...    ...   ...  \n",
       "c_2_30_3778    0.0    0.0    0.0   0.0  \n",
       "c_2_30_3808    0.0    0.0    0.0   0.0  \n",
       "c_2_30_3947    0.0    0.0    0.0   0.0  \n",
       "c_2_30_3987    0.0    0.0    0.0   0.0  \n",
       "c_2_30_4092    0.0    0.0    0.0   0.0  \n",
       "\n",
       "[1767 rows x 980 columns]"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# 基因表达矩阵：矩阵的每一行表示一个细胞，每一列表示一个基因。\n",
    "# cell x gene count n*m\n",
    "adata.to_df()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>0</th>\n",
       "      <th>1</th>\n",
       "      <th>2</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>MZT2A</th>\n",
       "      <td>-1.034772</td>\n",
       "      <td>-1.038119</td>\n",
       "      <td>1.027191</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>DUSP5</th>\n",
       "      <td>0.525707</td>\n",
       "      <td>0.871228</td>\n",
       "      <td>0.016519</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>MALAT1</th>\n",
       "      <td>-1.425137</td>\n",
       "      <td>-1.816856</td>\n",
       "      <td>1.324489</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CD74</th>\n",
       "      <td>2.380802</td>\n",
       "      <td>-2.017883</td>\n",
       "      <td>1.560988</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>COL9A2</th>\n",
       "      <td>-0.225960</td>\n",
       "      <td>1.865090</td>\n",
       "      <td>0.051355</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CASP8</th>\n",
       "      <td>1.764979</td>\n",
       "      <td>0.452462</td>\n",
       "      <td>-0.061717</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NR3C1</th>\n",
       "      <td>1.513145</td>\n",
       "      <td>-0.358040</td>\n",
       "      <td>-1.000590</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AZU1</th>\n",
       "      <td>-2.397397</td>\n",
       "      <td>-3.401837</td>\n",
       "      <td>0.922211</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>BEST1</th>\n",
       "      <td>-1.048677</td>\n",
       "      <td>1.237246</td>\n",
       "      <td>-0.163087</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NLRP1</th>\n",
       "      <td>-2.344598</td>\n",
       "      <td>-1.711320</td>\n",
       "      <td>0.088454</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>381 rows × 3 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "               0         1         2\n",
       "MZT2A  -1.034772 -1.038119  1.027191\n",
       "DUSP5   0.525707  0.871228  0.016519\n",
       "MALAT1 -1.425137 -1.816856  1.324489\n",
       "CD74    2.380802 -2.017883  1.560988\n",
       "COL9A2 -0.225960  1.865090  0.051355\n",
       "...          ...       ...       ...\n",
       "CASP8   1.764979  0.452462 -0.061717\n",
       "NR3C1   1.513145 -0.358040 -1.000590\n",
       "AZU1   -2.397397 -3.401837  0.922211\n",
       "BEST1  -1.048677  1.237246 -0.163087\n",
       "NLRP1  -2.344598 -1.711320  0.088454\n",
       "\n",
       "[381 rows x 3 columns]"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# 三维基因构象数据：矩阵的每一行表示一个基因，每一列表示一个空间维度（X、Y、Z）\n",
    "# conformation n*m'\n",
    "adata.uns['X']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>MZT2A</th>\n",
       "      <th>DUSP5</th>\n",
       "      <th>MALAT1</th>\n",
       "      <th>CD74</th>\n",
       "      <th>COL9A2</th>\n",
       "      <th>NEAT1</th>\n",
       "      <th>TYK2</th>\n",
       "      <th>HLA-B</th>\n",
       "      <th>PSAP</th>\n",
       "      <th>SRGN</th>\n",
       "      <th>...</th>\n",
       "      <th>LAMP2</th>\n",
       "      <th>CCL4L2</th>\n",
       "      <th>STAT5A</th>\n",
       "      <th>LIF</th>\n",
       "      <th>NANOG</th>\n",
       "      <th>CASP8</th>\n",
       "      <th>NR3C1</th>\n",
       "      <th>AZU1</th>\n",
       "      <th>BEST1</th>\n",
       "      <th>NLRP1</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>c_2_1_311</th>\n",
       "      <td>19.0</td>\n",
       "      <td>4.0</td>\n",
       "      <td>7.0</td>\n",
       "      <td>9.0</td>\n",
       "      <td>10.0</td>\n",
       "      <td>11.0</td>\n",
       "      <td>7.0</td>\n",
       "      <td>4.0</td>\n",
       "      <td>4.0</td>\n",
       "      <td>12.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>c_2_1_383</th>\n",
       "      <td>10.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>6.0</td>\n",
       "      <td>4.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>6.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>c_2_1_513</th>\n",
       "      <td>11.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>6.0</td>\n",
       "      <td>15.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>c_2_1_525</th>\n",
       "      <td>13.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>5.0</td>\n",
       "      <td>5.0</td>\n",
       "      <td>5.0</td>\n",
       "      <td>14.0</td>\n",
       "      <td>11.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>c_2_1_589</th>\n",
       "      <td>6.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>8.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>5.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>c_2_30_3778</th>\n",
       "      <td>5.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>c_2_30_3808</th>\n",
       "      <td>9.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>5.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>13.0</td>\n",
       "      <td>5.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>4.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>c_2_30_3947</th>\n",
       "      <td>10.0</td>\n",
       "      <td>4.0</td>\n",
       "      <td>15.0</td>\n",
       "      <td>8.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>6.0</td>\n",
       "      <td>4.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>4.0</td>\n",
       "      <td>...</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>c_2_30_3987</th>\n",
       "      <td>3.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>5.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>c_2_30_4092</th>\n",
       "      <td>2.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1767 rows × 381 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "             MZT2A  DUSP5  MALAT1  CD74  COL9A2  NEAT1  TYK2  HLA-B  PSAP  \\\n",
       "c_2_1_311     19.0    4.0     7.0   9.0    10.0   11.0   7.0    4.0   4.0   \n",
       "c_2_1_383     10.0    0.0     3.0   2.0     1.0    6.0   4.0    1.0   0.0   \n",
       "c_2_1_513     11.0    1.0     1.0   3.0     0.0    6.0  15.0    2.0   0.0   \n",
       "c_2_1_525     13.0    1.0     5.0   5.0     5.0   14.0  11.0    2.0   1.0   \n",
       "c_2_1_589      6.0    1.0     1.0   0.0     1.0    1.0   8.0    0.0   1.0   \n",
       "...            ...    ...     ...   ...     ...    ...   ...    ...   ...   \n",
       "c_2_30_3778    5.0    2.0     1.0   1.0     1.0    0.0   1.0    3.0   3.0   \n",
       "c_2_30_3808    9.0    2.0     5.0   1.0     0.0    3.0  13.0    5.0   0.0   \n",
       "c_2_30_3947   10.0    4.0    15.0   8.0     1.0    0.0   6.0    4.0   0.0   \n",
       "c_2_30_3987    3.0    1.0     5.0   3.0     0.0    2.0   3.0    0.0   0.0   \n",
       "c_2_30_4092    2.0    1.0     0.0   1.0     0.0    1.0   0.0    0.0   1.0   \n",
       "\n",
       "             SRGN  ...  LAMP2  CCL4L2  STAT5A  LIF  NANOG  CASP8  NR3C1  AZU1  \\\n",
       "c_2_1_311    12.0  ...    0.0     0.0     0.0  0.0    1.0    1.0    0.0   0.0   \n",
       "c_2_1_383     6.0  ...    0.0     0.0     0.0  0.0    0.0    0.0    0.0   0.0   \n",
       "c_2_1_513     1.0  ...    0.0     0.0     0.0  0.0    0.0    0.0    0.0   0.0   \n",
       "c_2_1_525     2.0  ...    0.0     0.0     0.0  0.0    0.0    0.0    0.0   0.0   \n",
       "c_2_1_589     5.0  ...    0.0     0.0     0.0  0.0    1.0    0.0    0.0   0.0   \n",
       "...           ...  ...    ...     ...     ...  ...    ...    ...    ...   ...   \n",
       "c_2_30_3778   0.0  ...    0.0     0.0     0.0  0.0    0.0    0.0    0.0   0.0   \n",
       "c_2_30_3808   4.0  ...    0.0     0.0     0.0  0.0    0.0    0.0    0.0   0.0   \n",
       "c_2_30_3947   4.0  ...    1.0     0.0     0.0  0.0    0.0    0.0    0.0   0.0   \n",
       "c_2_30_3987   1.0  ...    0.0     0.0     0.0  0.0    0.0    1.0    0.0   0.0   \n",
       "c_2_30_4092   0.0  ...    0.0     0.0     0.0  0.0    0.0    0.0    0.0   0.0   \n",
       "\n",
       "             BEST1  NLRP1  \n",
       "c_2_1_311      0.0    0.0  \n",
       "c_2_1_383      0.0    0.0  \n",
       "c_2_1_513      0.0    0.0  \n",
       "c_2_1_525      0.0    0.0  \n",
       "c_2_1_589      0.0    0.0  \n",
       "...            ...    ...  \n",
       "c_2_30_3778    0.0    0.0  \n",
       "c_2_30_3808    0.0    0.0  \n",
       "c_2_30_3947    1.0    0.0  \n",
       "c_2_30_3987    0.0    0.0  \n",
       "c_2_30_4092    0.0    1.0  \n",
       "\n",
       "[1767 rows x 381 columns]"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# 基因表达矩阵\n",
    "# cell x gene count n*m'\n",
    "adata.to_df()[adata.uns['X'].index]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 所有细胞共享的gene 3D conformation，获得gene-gene spatial covariance\n",
    "  - **基因的三维构象**\n",
    "    - 在细胞核内，基因并不是随机分布的，而是有特定的三维构象。这种构象决定了基因之间的空间距离。近距离的基因往往具有更高的共表达和共调控关系。\n",
    "\n",
    "  - **空间协方差**\n",
    "    - 空间协方差是指两个基因的表达值之间的依赖关系如何随着它们在三维空间中的距离变化。基于空间距离计算协方差矩阵，可以反映基因在空间中的关系。\n",
    "#### 步骤1：提取三维基因构象数据"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "3D Gene Conformation Data:\n",
      "               0         1         2\n",
      "MZT2A  -1.034772 -1.038119  1.027191\n",
      "DUSP5   0.525707  0.871228  0.016519\n",
      "MALAT1 -1.425137 -1.816856  1.324489\n",
      "CD74    2.380802 -2.017883  1.560988\n",
      "COL9A2 -0.225960  1.865090  0.051355\n",
      "...          ...       ...       ...\n",
      "CASP8   1.764979  0.452462 -0.061717\n",
      "NR3C1   1.513145 -0.358040 -1.000590\n",
      "AZU1   -2.397397 -3.401837  0.922211\n",
      "BEST1  -1.048677  1.237246 -0.163087\n",
      "NLRP1  -2.344598 -1.711320  0.088454\n",
      "\n",
      "[381 rows x 3 columns]\n"
     ]
    }
   ],
   "source": [
    "import numpy as np\n",
    "\n",
    "# 提取三维基因构象数据\n",
    "gene_3d_conformation = adata.uns['X']\n",
    "\n",
    "# 打印三维基因构象数据\n",
    "print(\"3D Gene Conformation Data:\")\n",
    "print(gene_3d_conformation)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### 步骤2：计算基因间的欧氏距离\n",
    "\n",
    "欧氏距离是计算两个点（在这里是两个基因在三维空间中的坐标）之间的直线距离。\n",
    "\n",
    "##### 欧氏距离公式\n",
    "\n",
    "对于两个点 $A (x_1, y_1, z_1)$ 和 $B (x_2, y_2, z_2)$，它们之间的欧氏距离计算公式为：\n",
    "$ d(A, B) = \\sqrt{(x_2 - x_1)^2 + (y_2 - y_1)^2 + (z_2 - z_1)^2} $\n",
    "\n",
    "- **`pdist`**: 这是SciPy库中的一个函数，用于计算成对距离。它接受一个数组，返回一个压缩的距离矩阵。\n",
    "- **`coordinates`**: 这是一个包含所有基因在三维空间中的坐标的数组，每行表示一个基因，每列表示一个坐标（X, Y, Z）。\n",
    "- **`metric='euclidean'`**: 这指定了距离度量方法为欧氏距离。欧氏距离是最常见的距离度量方法之一，表示两点之间的直线距离。"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[[0.         2.66498769 0.92043604 ... 2.73037211 2.56792464 1.74644571]\n",
      " [2.66498769 0.         3.5696464  ... 5.25584482 1.62631902 3.86178552]\n",
      " [0.92043604 3.5696464  0.         ... 1.90244078 3.4179143  1.5441268 ]\n",
      " ...\n",
      " [2.73037211 5.25584482 1.90244078 ... 0.         4.95156591 1.88567909]\n",
      " [2.56792464 1.62631902 3.4179143  ... 4.95156591 0.         3.23059221]\n",
      " [1.74644571 3.86178552 1.5441268  ... 1.88567909 3.23059221 0.        ]]\n"
     ]
    }
   ],
   "source": [
    "from scipy.spatial.distance import pdist, squareform\n",
    "\n",
    "# 提取三维坐标\n",
    "coordinates = gene_3d_conformation.values\n",
    "\n",
    "# 计算欧氏距离\n",
    "gene_distances = pdist(coordinates, metric='euclidean')\n",
    "\n",
    "# 转换为矩阵形式\n",
    "distance_matrix = squareform(gene_distances)\n",
    "\n",
    "# 打印距离矩阵\n",
    "print(distance_matrix)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### 步骤3：计算空间协方差矩阵\n",
    "\n",
    "**线性协方差函数 (Linear Covariance Function)**  \n",
    "  数学公式：\n",
    "  $ C(h) = \\sigma^2 \\left(1 - \\frac{h}{a}\\right), \\quad \\text{for} \\, h \\leq a $\n",
    "  $ C(h) = 0, \\quad \\text{for} \\, h > a $\n",
    "\n",
    "  解释：\n",
    "  线性协方差函数假设协方差随距离线性减小，直到距离达到阈值 $a$，此后协方差为零。该函数适用于短距离相关性较强的情况。\n",
    "\n",
    "**幂函数协方差函数 (Power Covariance Function)**  \n",
    "  数学公式：\n",
    "  $ C(h) = \\sigma^2 \\left(1 - \\left(\\frac{h}{a}\\right)^p\\right), \\quad \\text{for} \\, h \\leq a $\n",
    "  $ C(h) = 0, \\quad \\text{for} \\, h > a $\n",
    "\n",
    "  解释：\n",
    "  幂函数协方差函数使用幂次 $p$ 来控制协方差随距离的减小速度。当 $p$ 为 1 时，该函数变为线性函数。\n",
    "\n",
    "**高斯协方差函数 (Gaussian Covariance Function)**  \n",
    "  数学公式：\n",
    "  $ C(h) = \\sigma^2 \\exp\\left(-\\frac{h^2}{2a^2}\\right) $\n",
    "\n",
    "  解释：\n",
    "  高斯协方差函数假设协方差随距离的平方指数衰减，适用于具有平滑变化的空间现象。\n",
    "\n",
    "**球形协方差函数 (Spherical Covariance Function)**  \n",
    "  数学公式：\n",
    "  $ C(h) = \\sigma^2 \\left(1 - 1.5\\frac{h}{a} + 0.5\\left(\\frac{h}{a}\\right)^3\\right), \\quad \\text{for} \\, h \\leq a $\n",
    "  $ C(h) = 0, \\quad \\text{for} \\, h > a $\n",
    "\n",
    "  解释：\n",
    "  球形协方差函数在距离阈值 $a$ 内，协方差随距离非线性减小，且变化较为平缓，常用于地质和环境数据。\n",
    "\n",
    "**指数协方差函数 (Exponential Covariance Function)**  \n",
    "  数学公式：\n",
    "  $ C(h) = \\sigma^2 \\exp\\left(-\\frac{h}{a}\\right) $\n",
    "\n",
    "  解释：\n",
    "  指数协方差函数假设协方差随距离的指数减小，适用于具有非平滑变化的空间现象。\n",
    "\n",
    "**洞效应协方差函数 (Hole-effect Covariance Function)**  \n",
    "  数学公式：\n",
    "  $ C(h) = \\sigma^2 \\left(1 - c \\frac{\\sin(h/a)}{h/a}\\right) $\n",
    "\n",
    "  解释：\n",
    "  洞效应协方差函数模拟周期性变化的协方差，适用于具有周期性波动的空间现象。\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {},
   "outputs": [],
   "source": [
    "# 定义不同的协方差函数\n",
    "def linear_covariance(h, sigma2, a):\n",
    "    return sigma2 * (1 - h / a) * (h <= a)\n",
    "\n",
    "def power_covariance(h, sigma2, a, p):\n",
    "    return sigma2 * (1 - (h / a)**p) * (h <= a)\n",
    "\n",
    "def gaussian_covariance(h, sigma2, a):\n",
    "    return sigma2 * np.exp(-(h**2) / (2 * a**2))\n",
    "\n",
    "def spherical_covariance(h, sigma2, a):\n",
    "    return sigma2 * (1 - 1.5*(h/a) + 0.5*(h/a)**3) * (h <= a)\n",
    "\n",
    "def exponential_covariance(h, sigma2, a):\n",
    "    return sigma2 * np.exp(-h / a)\n",
    "\n",
    "def hole_effect_covariance(h, sigma2, a, c):\n",
    "    return sigma2 * (1 - c*np.sin(h/a) / (h/a))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 单个细胞中有部分的基因表达未知，使用covariance矩阵以及已知的基因表达值，impute未知基因表达值\n",
    "- 在单细胞RNA测序数据中，基因表达值为0可能有两种解释：\n",
    "\n",
    "  1. **真实的生物学现象**：在某些细胞中，某些基因的表达确实为0。这种情况表明该基因在该细胞中没有被转录。\n",
    "  2. **技术原因导致的缺失**：由于测序深度或技术限制，一些基因的表达值可能没有被检测到，尽管它们在实际中可能有非零表达。这种情况通常被称为“dropout”现象。\n",
    "   \n",
    "- **对0值进行处理**:  \n",
    "将这些值视为缺失值并进行插值处理。将0值替换为缺失值（NaN），然后进行Kriging插值\n",
    "\n",
    "#### 步骤1：存储基因表达矩阵\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "<class 'pandas.core.frame.DataFrame'>\n",
      "Index: 1767 entries, c_2_1_311 to c_2_30_4092\n",
      "Columns: 381 entries, MZT2A to NLRP1\n",
      "dtypes: float64(381)\n",
      "memory usage: 5.2+ MB\n",
      "None\n",
      "           MZT2A  DUSP5  MALAT1  CD74  COL9A2  NEAT1  TYK2  HLA-B  PSAP  SRGN  \\\n",
      "c_2_1_311   19.0    4.0     7.0   9.0    10.0   11.0   7.0    4.0   4.0  12.0   \n",
      "c_2_1_383   10.0    0.0     3.0   2.0     1.0    6.0   4.0    1.0   0.0   6.0   \n",
      "c_2_1_513   11.0    1.0     1.0   3.0     0.0    6.0  15.0    2.0   0.0   1.0   \n",
      "c_2_1_525   13.0    1.0     5.0   5.0     5.0   14.0  11.0    2.0   1.0   2.0   \n",
      "c_2_1_589    6.0    1.0     1.0   0.0     1.0    1.0   8.0    0.0   1.0   5.0   \n",
      "\n",
      "           ...  LAMP2  CCL4L2  STAT5A  LIF  NANOG  CASP8  NR3C1  AZU1  BEST1  \\\n",
      "c_2_1_311  ...    0.0     0.0     0.0  0.0    1.0    1.0    0.0   0.0    0.0   \n",
      "c_2_1_383  ...    0.0     0.0     0.0  0.0    0.0    0.0    0.0   0.0    0.0   \n",
      "c_2_1_513  ...    0.0     0.0     0.0  0.0    0.0    0.0    0.0   0.0    0.0   \n",
      "c_2_1_525  ...    0.0     0.0     0.0  0.0    0.0    0.0    0.0   0.0    0.0   \n",
      "c_2_1_589  ...    0.0     0.0     0.0  0.0    1.0    0.0    0.0   0.0    0.0   \n",
      "\n",
      "           NLRP1  \n",
      "c_2_1_311    0.0  \n",
      "c_2_1_383    0.0  \n",
      "c_2_1_513    0.0  \n",
      "c_2_1_525    0.0  \n",
      "c_2_1_589    0.0  \n",
      "\n",
      "[5 rows x 381 columns]\n"
     ]
    }
   ],
   "source": [
    "# 存储基因表达矩阵\n",
    "gene_expression_matrix = adata.to_df()[adata.uns['X'].index]\n",
    "\n",
    "# 查看基因表达矩阵的基本信息\n",
    "print(gene_expression_matrix.info())\n",
    "\n",
    "# 查看基因表达矩阵的前几行\n",
    "print(gene_expression_matrix.head())"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### 步骤2：将0值替换为NaN"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "将0值替换为NaN后的基因数量:\n",
      "MZT2A       33\n",
      "DUSP5       54\n",
      "MALAT1     109\n",
      "CD74       162\n",
      "COL9A2     203\n",
      "          ... \n",
      "CASP8     1590\n",
      "NR3C1     1590\n",
      "AZU1      1589\n",
      "BEST1     1587\n",
      "NLRP1     1589\n",
      "Length: 381, dtype: int64\n",
      "将0值替换为NaN后的细胞数量:\n",
      "c_2_1_311      250\n",
      "c_2_1_383      311\n",
      "c_2_1_513      300\n",
      "c_2_1_525      255\n",
      "c_2_1_589      316\n",
      "              ... \n",
      "c_2_30_3778    332\n",
      "c_2_30_3808    305\n",
      "c_2_30_3947    287\n",
      "c_2_30_3987    333\n",
      "c_2_30_4092    349\n",
      "Length: 1767, dtype: int64\n"
     ]
    }
   ],
   "source": [
    "# 将表达矩阵中的0值替换为NaN\n",
    "gene_expression_matrix_with_nan = gene_expression_matrix.replace(0, np.nan)\n",
    "\n",
    "# 检查替换后的缺失值情况\n",
    "missing_values_summary = gene_expression_matrix_with_nan.isnull().sum()\n",
    "print(\"将0值替换为NaN后的基因数量:\")\n",
    "print(missing_values_summary[missing_values_summary > 0])\n",
    "\n",
    "missing_cells_summary = gene_expression_matrix_with_nan.isnull().sum(axis=1)\n",
    "print(\"将0值替换为NaN后的细胞数量:\")\n",
    "print(missing_cells_summary[missing_cells_summary > 0])\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### 步骤3：选择一个有缺失值的细胞进行插值"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Cell c_2_1_311 Original Expression Values (with missing values):\n",
      "MZT2A     19.0\n",
      "DUSP5      4.0\n",
      "MALAT1     7.0\n",
      "CD74       9.0\n",
      "COL9A2    10.0\n",
      "          ... \n",
      "CASP8      1.0\n",
      "NR3C1      NaN\n",
      "AZU1       NaN\n",
      "BEST1      NaN\n",
      "NLRP1      NaN\n",
      "Name: c_2_1_311, Length: 381, dtype: float64\n"
     ]
    }
   ],
   "source": [
    "# 找到一个有缺失值的细胞\n",
    "missing_cells = missing_cells_summary[missing_cells_summary > 0]\n",
    "cell_index = missing_cells.index[0]  # 取第一个有缺失值的细胞作为示例\n",
    "\n",
    "# 提取这个细胞的基因表达数据\n",
    "cell_expression = gene_expression_matrix_with_nan.loc[cell_index]\n",
    "\n",
    "print(f\"Cell {cell_index} Original Expression Values (with missing values):\")\n",
    "print(cell_expression)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### 步骤4：进行Kriging插值\n",
    "克里金插值是一种利用已知点之间的空间相关性来预测未知点值的地统计学方法。协方差矩阵使用负指数函数来计算，这是为了反映点与点之间的空间相关性。\n",
    "\n",
    "1. **协方差矩阵 $ \\mathbf{K} $**:  \n",
    "    $ \\mathbf{K} $ 是已知点之间的协方差矩阵，其元素 $ K_{ij} $ 表示第 $ i $ 个已知点和第 $ j $ 个已知点之间的协方差。\n",
    "   \n",
    "2. **协方差向量 $ \\mathbf{k} $**:  \n",
    "    $ \\mathbf{k} $ 是未知点与已知点之间的协方差向量，其元素 $ k_i $ 表示未知点和第 $ i $ 个已知点之间的协方差。\n",
    "\n",
    "3. **已知值向量 $ \\mathbf{Z} $**:  \n",
    "    $ \\mathbf{Z} $ 是已知点的值组成的向量。\n",
    "\n",
    "4. **克里金公式推导**\n",
    "\n",
    "   克里金插值的基本公式为：\n",
    "      $ \\hat{Z}(u_0) = \\sum_{i=1}^{n} \\lambda_i Z(u_i) $\n",
    "\n",
    "   其中 $ \\lambda_i $ 是权重，通过以下线性方程组求得：\n",
    "      $ \\mathbf{K} \\mathbf{\\lambda} = \\mathbf{k} $\n",
    "\n",
    "   求解上述方程组可以得到：\n",
    "      $ \\mathbf{\\lambda} = \\mathbf{K}^{-1} \\mathbf{k} $\n",
    "\n",
    "   将 $ \\mathbf{\\lambda} $ 代入插值公式，可以得到：\n",
    "      $ \\hat{Z}(u_0) = \\mathbf{k}^T \\mathbf{K}^{-1} \\mathbf{Z} $\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [],
   "source": [
    "def kriging_interpolation(known_values, covariance_df):\n",
    "    known_indices = known_values.dropna().index\n",
    "    unknown_indices = known_values[known_values.isna()].index\n",
    "    \n",
    "    # 构建已知和未知的协方差矩阵\n",
    "    K = covariance_df.loc[known_indices, known_indices].values\n",
    "    k = covariance_df.loc[unknown_indices, known_indices].values\n",
    "    \n",
    "    # 已知值\n",
    "    known_values_array = known_values.dropna().values\n",
    "    \n",
    "    # 计算预测值\n",
    "    predicted_values = np.dot(np.dot(k, np.linalg.inv(K)), known_values_array)\n",
    "    \n",
    "    # 返回插值结果\n",
    "    result = known_values.copy()\n",
    "    result[unknown_indices] = predicted_values\n",
    "    return result\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Cell c_2_1_311 Original Expression Values (with missing values):\n",
      "MZT2A     19.0\n",
      "DUSP5      4.0\n",
      "MALAT1     7.0\n",
      "CD74       9.0\n",
      "COL9A2    10.0\n",
      "          ... \n",
      "CASP8      1.0\n",
      "NR3C1      NaN\n",
      "AZU1       NaN\n",
      "BEST1      NaN\n",
      "NLRP1      NaN\n",
      "Name: c_2_1_311, Length: 381, dtype: float64\n",
      "Cell c_2_1_311 Imputed Expression Values (after kriging):\n",
      "MZT2A     19.000000\n",
      "DUSP5      4.000000\n",
      "MALAT1     7.000000\n",
      "CD74       9.000000\n",
      "COL9A2    10.000000\n",
      "            ...    \n",
      "CASP8      1.000000\n",
      "NR3C1      1.117599\n",
      "AZU1       0.756612\n",
      "BEST1      3.252107\n",
      "NLRP1      1.155923\n",
      "Name: c_2_1_311, Length: 381, dtype: float64\n"
     ]
    }
   ],
   "source": [
    "# 找到一个有缺失值的细胞\n",
    "missing_cells_summary = gene_expression_matrix_with_nan.isnull().sum(axis=1)\n",
    "missing_cells = missing_cells_summary[missing_cells_summary > 0]\n",
    "cell_index = missing_cells.index[0]  # 取第一个有缺失值的细胞作为示例\n",
    "\n",
    "# 提取这个细胞的基因表达数据\n",
    "cell_expression = gene_expression_matrix_with_nan.loc[cell_index]\n",
    "\n",
    "print(f\"Cell {cell_index} Original Expression Values (with missing values):\")\n",
    "print(cell_expression)\n",
    "\n",
    "# 对选定的细胞进行插值\n",
    "imputed_expression = kriging_interpolation(cell_expression, covariance_df)\n",
    "\n",
    "print(f\"Cell {cell_index} Imputed Expression Values (after kriging):\")\n",
    "print(imputed_expression)\n"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": ".venv",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
